### ISSUE TITLE
Question about the loss function

### ISSUE THREAD

**Issue by lyn1874** [xu-ji/IIC#70]
Hi, thanks a lot for your great work! I find it really interesting! 

However, I have a question regarding the loss function. To understand what kind of probability distribution between original `(z)` and augmented image `(z')` can maximise the MI, I calculated the heatmap using your provided function `IID_loss` (I assume there are only two clusters). 

I understand that the second term `H(z|z')` encourage one-hot prediction, and we want `(z)`  and `(z')`  have a similar probability distribution and fall into the same cluster, but then shouldn't the MI being maximised on the top-left and bottom right corner instead of top-right and bottom-left corners? 
[mi_histmap.pdf](https://github.com/xu-ji/IIC/files/4844656/mi_histmap.pdf)

**Comment by xu-ji** [comment]
The typical heatmaps that maximise MI using a neural network predictor are indeed diagonal, top left to bottom right. Though, any heatmap with rows and columns with one non-zero element each will have high MI (which is about consistent predictability), maximal MI is achieved when that element is 1/num_classes. The specific diagonality where z is mapped to the same z' class is due to the visual consistency of the network, i.e. z and z' represent outputs of the same function, of semantically same images.

Optimising for the best heatmap without our neural network setup will not be constrained by visual consistency and therefore not necessarily result in top-left to bottom-right direction. 

Your map, if it was in a 2x2 format, would have near one-element rows/columns, because z=0 is predictably mapped to z'=1 and z=1 is predictably mapped to z'=0. So it is correct that this has a high MI value.

**Comment by lyn1874** [comment]
Thanks a lot for your response. I agree with what you said, but what still confuse me is that if the maximum MI appear on top right and bottom left in my heatmap, then doesn't it mean my original image and augmented version tend to belong to two clusters? (which I thought they are supposed to be constrained in one cluster)

**Comment by xu-ji** [comment]
Are you using my training scripts? What is the data? Or are you just optimising the heatmaps without a neural network?

**Comment by lyn1874** [comment]
No, I didn't use the training scripts or the data yet. I am just amazed by how well the MI loss term work and want to understand how the optimisation of these two terms H(z) and H(z|z') influence the probability distribution between original and augmented data.

**Comment by xu-ji** [comment]
Then your heatmap is correct, as I explained before

 > any heatmap with rows and columns with one non-zero element each will have high MI (which is about consistent predictability), maximal MI is achieved when that element is 1/num_classes

Without a constraint that the variables z and z' represent the same thing, which in our case is provided by the neural network and data setup, you cannot expect the top-left to bottom-right diagonal.

> doesn't it mean my original image and augmented version tend to belong to two clusters?
 
You are not using images or augmentation or a neural network predictor. You are optimizing the distribution of two unconstrained variables.

**Comment by jizongFox** [comment]
Hi @xu-ji, 
I happen to see your conversation and your reply. Could you please just confirm one thing in your reply?
You said `any heatmap with one-hot rows and columns will have high MI (which is about consistent predictability)`
it is not one heatmap with uniform row and columns marginal distributions and confident joint distribution will have the maximum MI?

**Comment by xu-ji** [comment]
Yes you're right, that was unclear. Maximum MI will be achieved by a heatmap with uniform marginals (1/C...1/C) and cells which are either 1/C or 0, where C is number of classes. I meant one-hot conditioned on the row/col, or within each row/col only 1 cell has value > 0. Have edited my replies to make that clearer.

The point is that in general it's not true the same z class has to map to the same z' class for maximal MI.

**Comment by lyn1874** [comment]
okay, I understand it now. Thanks a lot!
